### DISCUSSION TITLE
tar_make_clustermq freezes execution after a certain amount of time (or target runs)

### DISCUSSION THREAD

**Discussion by xinstein** [ropensci/targets#689] [Help]
## Prework

* [x] Read and agree to the [code of conduct](https://ropensci.org/code-of-conduct/) and [contributing guidelines](https://github.com/ropensci/targets/blob/main/CONTRIBUTING.md).
* [x] If there is [already a relevant issue](https://github.com/ropensci/targets/issues), whether open or closed, comment on the existing thread instead of posting a new issue.
* [ ] Post a [minimal reproducible example](https://www.tidyverse.org/help/) like [this one](https://github.com/ropensci/targets/issues/256#issuecomment-754229683) so the maintainer can troubleshoot the problems you identify. A reproducible example is:
    * [ ] **Runnable**: post enough R code and data so any onlooker can create the error on their own computer.
    * [ ] **Minimal**: reduce runtime wherever possible and remove complicated details that are irrelevant to the issue at hand.
    * [ ] **Readable**: format your code according to the [tidyverse style guide](https://style.tidyverse.org/).

## Description

I have a large amount of targets (several tounsands) and I use clustermq to execute them, but the execution seems not being able to complete itself: it freezes at some random targets, a simple ctrl-c + rerun will suffice to make it continue. I need to do this 2~3 times before the whole network is built.

When it freezes, no computational resources seem to be occupied or short, neither cpu nor memory.

I don't know how to recreate a runnable example for this problem, but maybe you'll have some advice for me?
- using future instead of clustermq is somewhat unacceptible, since env init of individual task takes considerable overhead and persistent-worker style of clustermq helps a lot about that.

## Benchmarks

It freezes, for like ever.

## Information that might be useful
this is my tar_options.
```
tar_option_set(
    tidy_eval = F,
    error = "abridge",
    garbage_collection = T,
    # memory = "persistent", retrieval = "main", storage = "main",
    memory = "transient", retrieval = "worker", storage = "worker",
    workspace_on_error = T
)
```

**Comment by wlandau** [comment]
Hard to say what could be causing the hangup. Would be great if you could share a simple example that reproduces the problem. What kind of cluster do you have (or are you using your local machine?) and how many workers are you creating? What values do you assign to the `clustermq.scheduler` and `clustermq.template` global options? 

One thing you could try is to use `clustermq::Q()` directly instead of `targets`: https://mschubert.github.io/clustermq/.

Could be related to https://github.com/mschubert/clustermq/issues/222 if `_targets.R` puts a lot of data into memory before starting the pipeline.

**Reply by xinstein** [reply]
my clustermq is local, the simplest setup:
```
clustermq.scheduler = 'multiprocess'
```
number of workers is 4.

I don't think targets is putting data into memory, memory footprint keeps the same as when the last successful target before the hangup finishes.

**Reply by xinstein** [reply]
The last outputs of targets looks something like
```
...
built target xxxx3
built target xxxx2
start target yyyy3
start target yyyy2
built target xxxx4
start target yyyy4
```

**Reply by xinstein** [reply]
Sorry it is extremely hard to mimic a runnable example, but I could try your advices, currently I'm learning how to use clustermq directly.

**Comment by xinstein** [comment]
I further increased my time range, now there are even more targets.
It has come to the point where it's not runnable.
Now when I run tar_make_clustermq, I see a lot of skipped targets (takes a while to print all out), then a few targets start, then the hangup happens. No more targets could be built.

~~The only workaround is abandon parallel and use tar_make~~
~~The problem is NOT clustermq specific, tar_make also hangs with the same pattern.~~

~~Should this be a bug then?~~

I need more testing, will come back later

**Reply by wlandau** [reply]
I missed the part you mentioned about `clustermq.scheduler = "multiprocess"`. Pretty sure that's an instance of https://github.com/mschubert/clustermq/discussions/266 and https://github.com/mschubert/clustermq/issues/272. Unless you are on Windows, you can use `clustermq.scheduler = "multicore"`  and it should work. Another alternative is to use `targets::tar_make_future()` with `future.callr`.

```{r}
# _targets.R
library(future.callr)
plan(callr)
list(
  tar_target(...),
  ...
)
```

**Reply by xinstein** [reply]
Thank you! I tried the fixed branch of clustermq and after a long time of testing, I haven't seen the issue reappear.
